### ISSUE TITLE
Reenable Java implementations

### ISSUE THREAD

**Issue by pitr-ch** [ruby-concurrency/concurrent-ruby#717]
All the Java blocking abstraction implementations had to be rolled back because the java threading is not compatible with ruby threading. E.g. a thread blocked by java CountDownLatch implementation cannot be killed from Ruby.

**Comment by pitr-ch** [comment]
cc @enebo @headius could JRuby's thread kill and other operations be made compatible with java interrupt mechanism to unblock Ruby threads blocked by Java blocking classes like CountDownLatch, Exchanger, etc.

**Comment by headius** [comment]
JRuby can't violate anything about Java's threads, so we won't forcibly kill a target thread. We do have mechanisms to allow you to block with a wakeup mechanism, however, which is how we do locking and sleeps and selects interruptibly. See how sleep works in JRuby for an example.

**Comment by pitr-ch** [comment]
yeah, I use the JRuby's interruptible in https://github.com/ruby-concurrency/concurrent-ruby/blob/master/ext/com/concurrent_ruby/ext/SynchronizationLibrary.java#L244 however that does not work afaik with `java.util.concurrent.CountDownLatch` which can be only interrupted with https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#interrupt-- which is not used by JRuby. So either 1) c-r needs to write another CountDownLatch java implementation which would be compatible with JRuby's interuptible thread waiting or 2) JRuby has to call interupt, which I have no idea if that could be done. Therefore my question is about 2) if it would be feasible. Sorry for unclarity.

**Comment by headius** [comment]
@pitr-ch I believe you want to use this method: https://github.com/jruby/jruby/blob/1e235478afd39f5b62dd26dd3b8677fbc7582011/core/src/main/java/org/jruby/RubyThread.java#L1530

RubyThread.executeTask allows you to register an "unblocker" implementation that can be used to do that thread interrupt. I use the same mechanism here: https://github.com/jruby/jruby/blob/807900a0478fd4d9a5509eb4ad1172a1a89a2c1e/core/src/main/java/org/jruby/ext/thread/Queue.java#L581-L588

**Comment by pitr-ch** [comment]
Ah great that looks simpler that I thought. I may be able to fix this without rollbacking the Java implementations for the upcoming patch release. Thanks.

**Comment by enebo** [comment]
@pitr-ch hey is there still a Fixnum type check in concurrent-ruby?  We have a 2.5 implementation now and we see:

```text
/home/enebo/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.0.5-java/lib/concurrent/atomic/java_count_down_latch.rb:12: warning: constant ::Fixnum is deprecated
```

(Sorry for not just PRing or opening an issue :) ).

**Comment by pitr-ch** [comment]
@enebo thanks, fixed
